### PR TITLE
Foundations Profile Modify

### PR DESCRIPTION
**Pull request by rcruzfreelance**

### CODE REVIEW

**Comment by rcruzfreelance** [comment]
Checking build failure

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Exceptions.Modify.cs [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+
+            var expectedProfileDependencyException =
+                new ProfileDependencyException(failedProfileStorageException);
+
+            this.dateTimeBrokerMock.Setup(broker =>
+                broker.GetCurrentDateTimeOffset())
+                    .Throws(sqlException);
+
+            // when
+            ValueTask<Profile> addProfileTask =
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Exceptions.Modify.cs:38]: modifyProfileTask

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Exceptions.Modify.cs [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+            this.dateTimeBrokerMock.Verify(broker =>
+                broker.GetCurrentDateTimeOffset(),
+                    Times.Once);
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.SelectProfileByIdAsync(randomProfile.Id),
+                    Times.Never);
+
+            this.loggingBrokerMock.Verify(broker =>
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Exceptions.Modify.cs:53]: Logging comes above storage. Check the VNO calls too.

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Exceptions.Modify.cs [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+            this.dateTimeBrokerMock.VerifyNoOtherCalls();
+            this.storageBrokerMock.VerifyNoOtherCalls();
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+        }
+
+        [Fact]
+        public async void ShouldThrowValidationExceptionOnModifyIfReferenceErrorOccursAndLogItAsync()
+        {
+            // given
+            Profile foreignKeyConflictedProfile = CreateRandomProfile();
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Exceptions.Modify.cs:71]: Keep it consistent by random and then assigning it to this variable as you have done for message.

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Exceptions.Modify.cs [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+            string randomMessage = GetRandomMessage();
+            string exceptionMessage = randomMessage;
+
+            var foreignKeyConstraintConflictException =
+                new ForeignKeyConstraintConflictException(exceptionMessage);
+
+            var invalidProfileReferenceException =
+                new InvalidProfileReferenceException(foreignKeyConstraintConflictException);
+
+            var ProfileDependencyValidationException =
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Exceptions.Modify.cs:81]: small p

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Exceptions.Modify.cs [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+        }
+
+        [Fact]
+        public async Task ShouldThrowDependencyExceptionOnModifyIfDatabaseUpdateExceptionOccursAndLogItAsync()
+        {
+            // given
+            Profile randomProfile = CreateRandomProfile();
+            var databaseUpdateException = new DbUpdateException();
+
+            var failedProfileException =
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Exceptions.Modify.cs:124]: failedProfileStorageException

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Logic.Modify.cs [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+                GetRandomDateTimeOffset();
+
+            Profile randomProfile =
+                CreateRandomProfile(randomDate);
+
+            Profile inputProfile =
+                randomProfile;
+
+            Profile storageProfile =
+                inputProfile.DeepClone();
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Logic.Modify.cs:32]: Why deepclone here?

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Logic.Modify.cs [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+            Profile updatedProfile =
+                inputProfile;
+
+            Profile expectedProfile =
+                updatedProfile.DeepClone();
+
+            Guid profileId =
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Logic.Modify.cs:40]: inputProfileId

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Logic.Modify.cs [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+namespace Taarafo.Core.Tests.Unit.Services.Foundations.Profiles
+{
+    public partial class ProfileServiceTests
+    {
+        [Fact]
+        public async Task ShouldModifyProfileAsync()
+        {
+            // given
+            DateTimeOffset randomDate =
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Logic.Modify.cs:22]: My thought - break the line only for 120 and belly patterns.

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+            // then
+            await Assert.ThrowsAsync<ProfileValidationException>(() =>
+                modifyProfileTask.AsTask());
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogError(It.Is(SameExceptionAs(
+                    expectedProfileValidationException))),
+                        Times.Once);
+
+            this.storageBrokerMock.Verify(broker =>
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs:40]: Why not insert verification?

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+                key: nameof(Profile.Username),
+                values: "Text is required");
+
+            invalidProfileException.AddData(
+                key: nameof(Profile.Email),
+                values: "Text is required");
+
+            invalidProfileException.AddData(
+                key: nameof(Profile.CreatedDate),
+                "Date is required",
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs:82]: values:

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+            invalidProfileException.AddData(
+                key: nameof(Profile.UpdatedDate),
+                values: "Date is required");
+
+            var expectedProfileValidationException =
+                new ProfileValidationException(invalidProfileException);
+
+            this.dateTimeBrokerMock.Setup(broker =>
+                broker.GetCurrentDateTimeOffset())
+                    .Returns(DateTimeOffset.UtcNow);
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs:94]: GetRandomDateTime

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+        [Theory]
+        [MemberData(nameof(MinutesBeforeOrAfter))]
+        public async Task ShouldThrowValidationExceptionOnModifyIfUpdatedDateIsNotRecentAndLogItAsync(int minutes)
+        {
+            // given
+            DateTimeOffset dateTime = GetRandomDateTimeOffset();
+            Profile randomProfile = CreateRandomProfile(dateTime);
+            Profile inputProfile = randomProfile;
+            inputProfile.CreatedDate = dateTime.AddMinutes(minutes);
+            var invalidProfileException =
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs:186]: new line break

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+                new InvalidProfileException();
+
+            invalidProfileException.AddData(
+                key: nameof(Profile.CreatedDate),
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs:190]: UpdatedDate

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+                values: "Date is not recent");
+
+            var expectedProfileValidatonException =
+                new ProfileValidationException(invalidProfileException);
+
+            this.dateTimeBrokerMock.Setup(broker =>
+                broker.GetCurrentDateTimeOffset())
+                .Returns(dateTime);
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs:198]: Tab

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs [hunk 15]
```diff
@@ -0,0 +1,278 @@
...
+            this.dateTimeBrokerMock.VerifyNoOtherCalls();
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+            this.storageBrokerMock.VerifyNoOtherCalls();
+        }
+
+        [Fact]
+        public async Task ShouldThrowValidationExceptionOnModifyIfProfileDoesNotExistAndLogItAsync()
+        {
+            // given
+            DateTimeOffset someDate = DateTimeOffset.UtcNow;
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs:230]: Random Date Time

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs [hunk 16]
```diff
@@ -0,0 +1,278 @@
...
+            Profile someProfile = CreateRandomProfile(someDate);
+
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs:232]: Remove this line

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs [hunk 17]
```diff
@@ -0,0 +1,278 @@
...
+            Profile nullProfile = null;
+
+            var notFoundProfileException =
+                new NotFoundProfileException(someProfile.Id);
+
+            var expectedProfileValidationException =
+                new ProfileValidationException(notFoundProfileException);
+
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs:240]: Remove this line

File: Taarafo.Core/Services/Foundations/Profiles/IProfileService.cs [hunk 18]
```diff
@@ -15,5 +15,6 @@ public partial interface IProfileService
         ValueTask<Profile> AddProfileAsync(Profile profile);
         IQueryable<Profile> RetrieveAllProfiles();
         ValueTask<Profile> RetrieveProfileByIdAsync(Guid profileId);
+        ValueTask<Profile> ModifyProfileAsync(Profile profile);
```
Comment (ShriHumrudha) [inline Taarafo.Core/Services/Foundations/Profiles/IProfileService.cs:18]: Order the CRUD opetation

File: Taarafo.Core/Services/Foundations/Profiles/ProfileService.Exceptions.cs [hunk 19]
```diff
@@ -59,6 +59,15 @@ private async ValueTask<Profile> TryCatch(ReturningProfileFunction returningProf
 
                 throw CreateAndLogDependencyValidationException(invalidProfileReferenceException);
             }
+
```
Comment (ShriHumrudha) [inline Taarafo.Core/Services/Foundations/Profiles/ProfileService.Exceptions.cs:62]: extra line

File: Taarafo.Core/Services/Foundations/Profiles/ProfileService.Exceptions.cs [hunk 20]
```diff
@@ -59,6 +59,15 @@ private async ValueTask<Profile> TryCatch(ReturningProfileFunction returningProf
...
+            catch (DbUpdateConcurrencyException databaseUpdateConcurrencyException)
+            {
+                var lockedProfileException =
+                    new LockedProfileException(databaseUpdateConcurrencyException);
+
+                throw CreateAndLogDependencyValidationException(lockedProfileException);
+            }
+
```
Comment (ShriHumrudha) [inline Taarafo.Core/Services/Foundations/Profiles/ProfileService.Exceptions.cs:70]: extra line

File: Taarafo.Core/Services/Foundations/Profiles/ProfileService.Validations.cs [hunk 21]
```diff
@@ -32,6 +32,27 @@ private void ValidateProfileOnAdd(Profile profile)
                 (Rule: IsNotRecent(profile.CreatedDate), Parameter: nameof(Profile.CreatedDate)));
         }
 
+        private void ValidateProfileOnModify(Profile profile)
+        {
+            ValidateProfileIsNotNull(profile);
+
+            Validate(
+                (Rule: IsInvalid(profile.Id), Parameter: nameof(Profile.Id)),
```
Comment (ShriHumrudha) [inline Taarafo.Core/Services/Foundations/Profiles/ProfileService.Validations.cs:40]: parameter aliasning?

**Review by ShriHumrudha — CHANGES_REQUESTED** [review]

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs (hunk 13 above)
Comment (rcruzfreelance) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs:190]: done

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs (hunk 12 above)
Comment (rcruzfreelance) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs:186]: done

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs (hunk 11 above)
Comment (rcruzfreelance) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs:94]: done

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs (hunk 10 above)
Comment (rcruzfreelance) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs:82]: done

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs (hunk 9 above)
Comment (rcruzfreelance) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs:40]: done

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Logic.Modify.cs (hunk 8 above)
Comment (rcruzfreelance) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Logic.Modify.cs:22]: done

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Logic.Modify.cs (hunk 7 above)
Comment (rcruzfreelance) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Logic.Modify.cs:40]: done

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Logic.Modify.cs (hunk 6 above)
Comment (rcruzfreelance) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Logic.Modify.cs:32]: done

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Exceptions.Modify.cs (hunk 5 above)
Comment (rcruzfreelance) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Exceptions.Modify.cs:124]: done

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Exceptions.Modify.cs (hunk 4 above)
Comment (rcruzfreelance) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Exceptions.Modify.cs:81]: done

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Exceptions.Modify.cs (hunk 3 above)
Comment (rcruzfreelance) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Exceptions.Modify.cs:71]: done

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Exceptions.Modify.cs (hunk 2 above)
Comment (rcruzfreelance) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Exceptions.Modify.cs:53]: done

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Exceptions.Modify.cs (hunk 1 above)
Comment (rcruzfreelance) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Exceptions.Modify.cs:38]: done

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs (hunk 14 above)
Comment (rcruzfreelance) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs:198]: done

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs (hunk 15 above)
Comment (rcruzfreelance) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs:230]: done

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs (hunk 16 above)
Comment (rcruzfreelance) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs:232]: done

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs (hunk 17 above)
Comment (rcruzfreelance) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs:240]: done

File: Taarafo.Core/Services/Foundations/Profiles/IProfileService.cs (hunk 18 above)
Comment (rcruzfreelance) [reply Taarafo.Core/Services/Foundations/Profiles/IProfileService.cs:18]: done I switched the RetrieveAll with the RetrieveById

File: Taarafo.Core/Services/Foundations/Profiles/ProfileService.Exceptions.cs (hunk 19 above)
Comment (rcruzfreelance) [reply Taarafo.Core/Services/Foundations/Profiles/ProfileService.Exceptions.cs:62]: done

File: Taarafo.Core/Services/Foundations/Profiles/ProfileService.Exceptions.cs (hunk 20 above)
Comment (rcruzfreelance) [reply Taarafo.Core/Services/Foundations/Profiles/ProfileService.Exceptions.cs:70]: done

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs [hunk 22]
```diff
@@ -0,0 +1,278 @@
...
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+            this.dateTimeBrokerMock.VerifyNoOtherCalls();
+            this.storageBrokerMock.VerifyNoOtherCalls();
+        }
+
+        [Fact]
+        public async Task ShouldThrowValidationExceptionOnModifyIfUpdatedDateIsNotSameAsCreatedDateAndLogItAsync()
+        {
+            // given
+            DateTimeOffset randomDateTime =
```
[the remark this answers is not in the document]
Comment (rcruzfreelance) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Validations.Modify.cs:126]: done, I used the Comment Service unit test for modify as an example

File: Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Exceptions.Modify.cs (hunk 1 above)
Comment (rcruzfreelance) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Profiles/ProfileServiceTests.Exceptions.Modify.cs]: done

**Pull request merged**